guides/introduction/Embedded Schemas.md

```diff
@@ -92,57 +89,107 @@ end
   # ...
 
-  def changeset(profile, attrs \\ %{}) do
+  def changeset(%UserProfile{} = profile, attrs \\ %{}) do
     profile
-    |> cast(attrs, [:age, :favorite_color, :avatar_url])
-    |> validate_required([:age, :favorite_color])
+    |> cast(attrs, [:online, :dark_mode, :visibility])
+    |> validate_required([:online, :visibility])
   end
 end
+
+profile = %UserProfile{}
+UserProfile.changeset(profile, %{online: true, visibility: :public})
 ```
 
-On the user side, you also define a `changeset/2` function, and then you use `cast_embed/3` to invoke the `UserProfile` changeset:
+Meanwhile, the User changeset function can require it's own validations without worrying about the details of the UserProfile changes because it can pass that responsibility to UserProfile via `cast_embed/3`. A validation failure in an embed will cause the parent changeset to be invalid, even if the parent changeset itself had no errors.
 
 ```elixir
 defmodule User do
-  use Ecto.Schema
-
   # ...
 
   def changeset(user, attrs \\ %{}) do
+    user
+    |> cast(attrs, [:full_name, :email, :avatar_url])
+    |> cast_embed(:profile, required: true)
+  end
+end
+
+changeset = User.changeset(%User{}, %{profile: %{online: true}})
+changeset.valid? # => false; "visibility can't be blank"
+changeset = User.changeset(%User{}, %{profile: %{online: true, visibility: :public}})
+changeset.valid? # => true
+```
+
+In situations where you have kept the embedded schema within the parent module, e.g., you have not extracted a UserProfile, you can still have custom changeset functions for the embedded data within the parent schema.
+
+```elixir
+defmodule User do
+  use Ecto.Schema
+
+  schema "users" do
+    field :full_name, :string
+    field :email, :string
+    field :avatar_url, :string
+    field :confirmed_at, :naive_datetime
+
+    embeds_one :profile, Profile do
+      field :online, :boolean
+      field :dark_mode, :boolean
+      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
+    end
+
+    timestamps()
+  end
+
+  def changeset(%User{} = user, attrs \\ %{}) do
     user
     |> cast(attrs, [:full_name, :email])
-    # By default it calls UserProfile.changeset/2, pass the :with option to change it
-    |> cast_embed(:user_profile, required: true)
+    |> cast_embed(:profile, required: true, with: &profile_changeset/2)
+  end
+
+  def profile_changeset(profile, attrs \\ %{}) do
+    profile
+    |> cast(attrs, [:online, :dark_mode, :visibility])
+    |> validate_required([:online, :visibility])
   end
 end
+
+changeset = User.changeset(%User{}, %{profile: %{online: true, visibility: :public}})
+changeset.valid? # => true
+```
+
+### Querying embedded data
+
+Once you have written embedded data to the database, you can use it in queries on the parent struct.
+
+<!-- TODO: Actually proves this works with local test data. Otherwise, user JSON query fragments -->
+```elixir
+from u in User, where: u.profile.dark_mode == true
```

Comment:
I am not sure what to say. The above should just work on more recent Ecto versions. What version you are using?

Reply:
@josevalim sorry I tried to get it working but can't. Here's the error:

```
** (Ecto.Query.CompileError) cannot fetch field `online` from `p.profile`. Can only fetch fields from:

  * sources, such as `p` in `from p in Post`
  * named bindings, such as `as(:post)` in `from Post, as: :post`
  * parent named bindings, such as `parent_as(:post)` in a subquery

    (ecto 3.7.1) expanding macro: Ecto.Query.where/3
    iex:19: (file)
    (ecto 3.7.1) expanding macro: Ecto.Query.from/2
    iex:19: (file)
```

Here's some links that led me to believe this is not possible:

* https://groups.google.com/g/elixir-ecto/c/Z20aGxm3IhE
* https://stackoverflow.com/questions/63930941/queriyng-embedded-schema-without-using-fragment

Reply:
Correct. I got the syntax messed up. You should do this: `p.profile["dark_mode"]`.

Reply:
Ecto.Query.from is not converting to the correct jsonpath syntax, it seems. In the code below first query is healthy, second one fails because it's converting to jsonpath using `#>` instead of `##>` which would allow for the text comparison to succeed.

```
iex(22)> (Ecto.Query.from p in Person, where: p.age == 12) |> Repo.all

07:58:48.330 [debug] QUERY OK source="people" db=1.8ms idle=1095.6ms
SELECT p0."id", p0."first_name", p0."last_name", p0."age", p0."profile" FROM "people" AS p0 WHERE (p0."age" = 12) []
[
  %Friends.Person{
    __meta__: #Ecto.Schema.Metadata<:loaded, "people">,
    age: 12,
    first_name: nil,
    id: 1,
    last_name: nil,
    profile: %Friends.Person.Profile{
      dark_mode: nil,
      id: "b7c1fdef-bdfe-4303-8dfa-830ae60c1fe5",
      online: "yeah"
    }
  }
]
iex(23)> (Ecto.Query.from p in Person, where: p.profile["online"] == "yeah") |> Repo.all

07:59:06.929 [debug] QUERY ERROR source="people" db=0.0ms queue=1.3ms idle=1695.6ms
SELECT p0."id", p0."first_name", p0."last_name", p0."age", p0."profile" FROM "people" AS p0 WHERE ((p0."profile"#>'{"online"}') = 'yeah') []
** (Postgrex.Error) ERROR 22P02 (invalid_text_representation) invalid input syntax for type json

    query: SELECT p0."id", p0."first_name", p0."last_name", p0."age", p0."profile" FROM "people" AS p0 WHERE ((p0."profile"#>'{"online"}') = 'yeah')

Token "yeah" is invalid.
    (ecto_sql 3.7.1) lib/ecto/adapters/sql.ex:760: Ecto.Adapters.SQL.raise_sql_call_error/1
    (ecto_sql 3.7.1) lib/ecto/adapters/sql.ex:693: Ecto.Adapters.SQL.execute/5
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:219: Ecto.Repo.Queryable.execute/4
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
```

I can generates this same error in psql:

```
tecto_repo=# select profile from people where profile #>> '{"online"}' = 'yeah';
                                       profile                                       
-------------------------------------------------------------------------------------
 {"id": "b7c1fdef-bdfe-4303-8dfa-830ae60c1fe5", "online": "yeah", "dark_mode": null}
(1 row)

tecto_repo=# select profile from people where profile #> '{"online"}' = 'yeah';
ERROR:  invalid input syntax for type json
LINE 1: ... profile from people where profile #> '{"online"}' = 'yeah';
                                                                ^
DETAIL:  Token "yeah" is invalid.
CONTEXT:  JSON data, line 1: yeah
```

Maybe it's a Postgrex issue, or maybe I'm missing something. I can't find if there's documentation on this feature. (Maybe I found the next guide to write, XD).

Reply:
How is your migration? Did you define the field as :json or :jsonb?

Reply:
```
tecto_repo=# \d people
                                      Table "public.people"
   Column   |          Type          | Collation | Nullable |              Default               
------------+------------------------+-----------+----------+------------------------------------
 id         | bigint                 |           | not null | nextval('people_id_seq'::regclass)
 first_name | character varying(255) |           |          | 
 last_name  | character varying(255) |           |          | 
 age        | integer                |           |          | 
 profile    | jsonb                  |           |          | 
Indexes:
    "people_pkey" PRIMARY KEY, btree (id)
```

Reply:
```elixir
    alter table :people do
      add :profile, :map
    end
```